handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/move/ProductMoveEndpoint.scala

```
@@ -49,46 +49,52 @@ object ProductMoveEndpoint {
         )
     endpointDescription
-      .serverLogic[TIO] { (_, input) => run(input).tapEither(result => ZIO.log("result tapped: " + result)).map(Right.apply) }
+      .serverLogic[TIO] { (subscriptionName, postData) => run(subscriptionName, postData).tapEither(result => ZIO.log("result tapped: " + result)).map(Right.apply) }
   }
-  private def run(input: ExpectedInput): TIO[OutputBody] =
-    runWithEnvironment(input).provide(
+
+  private def run(subscriptionName: String, postData: ExpectedInput): TIO[OutputBody] =
+    productMove(subscriptionName, postData).provide(
+      SubscribeLive.layer,
+      GetSubscriptionLive.layer,
+      CancellationLive.layer,
       AwsS3Live.layer,
       AwsCredentialsLive.layer,
       SttpClientLive.layer,
       ZuoraClientLive.layer,
-      GetSubscriptionLive.layer,
       ZuoraGetLive.layer,
       GuStageLive.layer,
     )
 
-  private[productmove] def runWithEnvironment(postData: ExpectedInput): ZIO[GetSubscription, String, OutputBody] =
+  private[productmove] def productMove(subscriptionName: String, postData: ExpectedInput): ZIO[GetSubscription with Subscribe with Cancellation, String, OutputBody] =
     for {
       _ <- ZIO.log("PostData: " + postData.toString)
-      sub <- GetSubscription.get(if (postData.targetProductId == "true") "A-S00090478" else "A-S00339056") //DEV - for testing locally
-      _ <- ZIO.log("Sub: " + sub.toString)
-    } yield Success("asdf",
-      MoveToProduct(
-        id = "123",
-        name = "Digital Pack",
-        billing = Billing(
-          amount = Some(1199),
-          percentage = None,
-          currency = Currency.GBP,
-          frequency = Some(TimePeriod(TimeUnit.month, 1)),
+      subscription <- GetSubscription.get(subscriptionName)
+
+      chargedThroughDate <- ZIO.fromOption(subscription.ratePlans.head.ratePlanCharges.head.chargedThroughDate).orElseFail(s"chargedThroughDate is null for subscription $subscriptionName.")
```

the chargedThroughDate comes through as null if the person is in a free trial (for a digital subscription) and has not made payment yet. Although we don't need to consider this in the MVP just thought I'd put it in as it was quick to do.

---

`.head` will throw an exception if there's no head, and I'm not sure how fromOption or flatMap would deal with that situation.
If we use `headOption` at each stage and flatMap it into the for comprehension then we can be sure that the error will be handled properly with an explicit message (as even if the exception is visible, it won't be clear which "head" failed from just the stack trace)

As a note, there will only be one (main) rate plan with one charge for contribution, but we need to beware for other products in future.  For example discounts would appear as a separate rate plan with a suitable discount charge.

---

for `.head`, will there for always be at least one rateplan if the user can get to the cancellation page? Or are we just making it a bit more robust and future proof?

 As for multiple rateplans, yep, we need to take that into account for future products, will make a note of it.

---

as discussed in the pairing earlier, it would be good to "fail fast" if there are the wrong number of rate plans, just in case this code is reused elsewhere.  Otherwise it might silently pick the wrong one.